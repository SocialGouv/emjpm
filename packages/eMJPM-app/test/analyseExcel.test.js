import {
  cleanInputData,
  isValidCodePostal,
  validateData
} from "../src/components/common/AnalyseExcel";

const VALID_CPS = ["13000", "89100", /*"97121", "97216",*/ "20000"];

describe("isValidCodePostal", () => {
  test("75000 should be invalid ", () => {
    expect(isValidCodePostal(75000)).toBe(false);
  });
  VALID_CPS.forEach(cp => {
    test(`${cp} should be valid`, () => {
      expect(isValidCodePostal(cp)).toBe(true);
    });
  });
});

// mock XLSX.utils.sheet_to_json
const jsonSheet = [
  [
    "date_ouverture",
    "type",
    null,
    "code_postal",
    "ville",
    "civilite",
    null,
    "annee",
    "numero_dossier",
    "residence",
    null,
    "adad",
    "mandataire_id"
  ],
  [
    43041,
    "tutelle",
    2,
    57100,
    "thionville",
    "n",
    null,
    4150,
    "RG20136Q",
    "à domicile",
    null,
    "q",
    10
  ],
  [
    42707,
    "curatelle",
    12,
    75005,
    "paris",
    "n",
    125,
    4000,
    "1236HPLY",
    "en établissement",
    null,
    "s",
    10
  ],
  [
    40395,
    "sauvegarde de justice ",
    52,
    85000,
    "marseille",
    "n",
    54,
    4500,
    "meourg120",
    "à domicile",
    null,
    "t",
    10
  ]
];

test("cleanInputData normalize input", () => {
  expect(cleanInputData(jsonSheet)).toMatchSnapshot();
});

const firstRow = {
  annee: null,
  civilite: null,
  code_postal: null,
  date_ouverture: null,
  numero_dossier: null,
  residence: null,
  type: null,
  ville: null
};

const tests = [
  {
    title: "should accept perfect valid data",
    success: true,
    data: [
      firstRow,
      {
        annee: 1992,
        civilite: "H",
        code_postal: 57100,
        date_ouverture: "02/11/2017",
        numero_dossier: "RG20136Q",
        residence: "A Domicile",
        type: "Tutelle",
        ville: "thionville"
      }
    ]
  },
  {
    title: "should not accept missing columns",
    success: false,
    data: [
      {
        annee: 1992,
        civilite: "H",
        code_postal: 57100,
        date_ouverture: "02/11/2017",
        numero_dossier: "RG20136Q",
        type: "Tutelle",
        ville: "thionville"
      },
      {
        annee: 1992,
        civilite: "H",
        code_postal: 57100,
        date_ouverture: "02/11/2017",
        residence: "A Domicile",
        numero_dossier: "RG20136Q",
        type: "Tutelle",
        ville: "thionville"
      }
    ]
  },
  {
    title: "should not accept invalid reference data",
    success: false,
    data: [
      firstRow,
      {
        annee: 1992,
        civilite: "H",
        code_postal: 57100,
        date_ouverture: "02/11/2017",
        numero_dossier: "RG20136Q",
        residence: "domicile",
        type: "Tutelle",
        ville: "thionville"
      }
    ]
  }
];

describe("validateData", () => {
  tests.forEach(t => {
    test(t.title, () => {
      const result = validateData(t.data);
      if (t.success) {
        expect(result.errors).toBeUndefined();
      } else {
        expect(result.errors).not.toBeUndefined();
      }
    });
  });
});
