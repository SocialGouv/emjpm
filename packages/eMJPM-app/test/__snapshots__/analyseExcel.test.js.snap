// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanInputData normalize input 1`] = `
Array [
  Object {
    "annee": 4150,
    "civilite": "n",
    "code_postal": 57100,
    "date_ouverture": "02/11/2017",
    "numero_dossier": "RG20136Q",
    "residence": "A Domicile",
    "type": "Tutelle",
    "ville": "thionville",
  },
  Object {
    "annee": 4000,
    "civilite": "n",
    "code_postal": 75005,
    "date_ouverture": "03/12/2016",
    "numero_dossier": "1236HPLY",
    "residence": "En Ã©tablissement",
    "type": "Curatelle",
    "ville": "paris",
  },
  Object {
    "annee": 4500,
    "civilite": "n",
    "code_postal": 85000,
    "date_ouverture": "05/08/2010",
    "numero_dossier": "meourg120",
    "residence": "A Domicile",
    "type": "Sauvegarde de justice",
    "ville": "marseille",
  },
]
`;
